lib/rucio/daemons/reaper/reaper.py

```diff
@@ -217,6 +222,26 @@ def delete_from_storage(heartbeat_handler, hb_payload, replicas, prot, rse_info,
                 deletion_dict['duration'] = duration
                 add_message('deletion-failed', deletion_dict)
 
+            finally:
+                # this control loop will run indefinetely until all deletions have gone through (or failed).
+                # it assumes that for each deletion a timeout will occur (which will fail the deletion).
+                # if that assumption is not true, we need to introduce a maximum retry counter to avoid the worker hanging
+                # on individual deletions.
+                num_replicas_processed += 1
+
+                # after each replica is deleted we evaluate how much time we have left to delete.
+                # if we are not able to delete all the replicas we got in, other workers will take the replicas because their update time will
+                # be more than 10 minutes (refer to function list_and_mark_unlocked_replicas).
+                # After 8 minutes, and if we still have some replicas to process, we bump the replicas updated_at field to current time (with will delay the selectability
+                # by other workers by 8+ minutes).
```

Comment:
I'd rephrase it like this for clarity:

```suggestion
                # Every 8 minutes, if there are still replicas to process,
                # we bump the updated_at field of the unprocessed replicas to current time,
                # which will prevent other workers from selecting
                # the unprocessed replicas for the next 8+ minutes
```